pkg/distributor/distributor.go

```diff
@@ -1205,26 +1204,34 @@ func (d *Distributor) push(ctx context.Context, pushReq *push.Request) (*mimirpb
 		localCtx = opentracing.ContextWithSpan(localCtx, sp)
 	}
 
-	keys := append(seriesKeys, metadataKeys...)
+	// All tokens, stored in order: series, metadata, ephemeral series.
+	keys := make([]uint32, len(seriesKeys)+len(metadataKeys)+len(ephemeralSeriesKeys))
 	initialMetadataIndex := len(seriesKeys)
+	initialEphemeralIndex := initialMetadataIndex + len(metadataKeys)
+	copy(keys, seriesKeys)
+	copy(keys[initialMetadataIndex:], metadataKeys)
+	copy(keys[initialEphemeralIndex:], ephemeralSeriesKeys)
 
 	// we must not re-use buffers now until all DoBatch goroutines have finished,
 	// so set this flag false and pass cleanup() to DoBatch.
 	cleanupInDefer = false
 
 	err = ring.DoBatch(ctx, ring.WriteNoExtend, subRing, keys, func(ingester ring.InstanceDesc, indexes []int) error {
 		timeseries := make([]mimirpb.PreallocTimeseries, 0, len(indexes))
```

Comment:
Can this improved? We actually iterate `indexes` first to count the number of items there, then preallocate `timeseries`, `ephemeral` and `metadata` accordingly.

Reply:
Good idea. Will send PR shortly.

Reply:
Done here: https://github.com/grafana/mimir/pull/3916